### DISCUSSION TITLE
Implement a persistently ordered list of info hashes to be consumed by the index/API (for DashMap torrent repository)

### DISCUSSION THREAD

**Discussion by mickvandijke** [torrust/torrust-tracker#788] [Proposals]
### UPDATE by @josecelano 

Relates to: https://github.com/torrust/torrust-tracker/pull/784

The [DashMap torrent repository implementation](https://github.com/torrust/torrust-tracker/blob/develop/packages/torrent-repository/src/repository/dash_map_mutex_std.rs) can only be used in production if it keeps torrents ordered.

For now, it's only used for benchmarking. In order to use it in production this should be implemented.

**Comment by josecelano** [comment]
Relates to: https://github.com/torrust/torrust-tracker/issues/565

The new tracker repository implementation does not keep order for torrents so the API would return random torrents. We need to keep the order in a different (new) data structure.

**Comment by mickvandijke** [comment]
I think we need to revisit if we actually need the tracker API to have the `api/v1/torrents` endpoint.

In the last meeting we decided that we didn't want a breaking change by merging [this](https://github.com/torrust/torrust-tracker/pull/645) PR that implements the new DashMap torrent repository. As it would break the ordering of torrents, which was supposedly needed for the Torrust index when it calls this api endpoint with pagination and such.

### Why do we need it
After looking into the index. I could actually not find any place where the tracker api endpoint `api/v1/torrents` is called. I also do not remember ever needing it when I worked on the index. I can only recall the index keeping track of its own torrents and then occasionally trying to fetch seeders/leechers information from the tracker per info hash. 

Am I missing something? Do you know when the index would have to go through all the tracker torrents @josecelano ?

If the feature is not used by the tracker nor the index, shouldn't we just drop it?

### How it currently works
Currently we keep all torrents in a `BTreeMap`, so that the torrents are ordered by key (info hash). When someone calls the get torrents api endpoint, they can specify a page limit and page number to retrieve all torrents using multiple calls.

However, even in the current implementation the torrent order is not always the same in between calls to the `get torrents` endpoint. If `remove_peerless_torrents` is enabled, empty torrents will automatically be dropped from the `BTreeMap`. So this will break the pagination.

### How it should work
If we somehow still need this feature for anything (could actually still be useful for testing), I think we should just return the entire detailed torrents map in one go. Yes it will be a big response if the torrent map is huge, but I don't see any other way to safely return all torrents. Meaning without any missing ones. The torrent map is just too dynamic. Even without sharding.

Thoughts @josecelano @da2ce7 ?

**Comment by josecelano** [comment]
> I think we need to revisit if we actually need the tracker API to have the `api/v1/torrents` endpoint.
> 
> In the last meeting we decided that we didn't want a breaking change by merging [this](https://github.com/torrust/torrust-tracker/pull/645) PR that implements the new DashMap torrent repository. As it would break the ordering of torrents, which was supposedly needed for the Torrust index when it calls this api endpoint with pagination and such.
> 
> ### Why do we need it
> After looking into the index. I could actually not find any place where the tracker api endpoint `api/v1/torrents` is called. I also do not remember ever needing it when I worked on the index. I can only recall the index keeping track of its own torrents and then occasionally trying to fetch seeders/leechers information from the tracker per info hash.
> 
> Am I missing something? Do you know when the index would have to go through all the tracker torrents @josecelano ?
> 
> If the feature is not used by the tracker nor the index, shouldn't we just drop it?
> 
> ### How it currently works
> Currently we keep all torrents in a `BTreeMap`, so that the torrents are ordered by key (info hash). When someone calls the get torrents api endpoint, they can specify a page limit and page number to retrieve all torrents using multiple calls.
> 
> However, even in the current implementation the torrent order is not always the same in between calls to the `get torrents` endpoint. If `remove_peerless_torrents` is enabled, empty torrents will automatically be dropped from the `BTreeMap`. So this will break the pagination.
> 
> ### How it should work
> If we somehow still need this feature for anything (could actually still be useful for testing), I think we should just return the entire detailed torrents map in one go. Yes it will be a big response if the torrent map is huge, but I don't see any other way to safely return all torrents. Meaning without any missing ones. The torrent map is just too dynamic. Even without sharding.
> 
> Thoughts @josecelano @da2ce7 ?

Hi @WarmBeer I think you are right. That endpoint is not used in the Index for the time being.
We loop over all torrents in the Index and then we use the endpoint to fetch a single torrent info. In fact, that's a process [I would like to review](https://github.com/torrust/torrust-index/discussions/428).

However, I think it's a pretty useful endpoint for testing. I use it a lot. Besides, the API is supposed to be generic (for purposes we don't even know), not only consumed by the Index. So it should be designed to fulfill most of the requirements other users might have. And I think that's an important endpoint. I don't see any problem in returning a partial view of the data. I guess that's what always happens when you do pagination. If you want we could add a `snapshot` endpoint or something like that which can give you the full data, but probably that should be an asynchronous task. I mean, you have to ask for that "report" and then you receive a notification when it's ready and finally you download it.

Regarding the option `remove_peerless_torrents`, I would remove that option. I would always remove peerless torrents from memory. See https://github.com/torrust/torrust-tracker/issues/566#issuecomment-1900105745. If we want to store all torrents I would do it similar way to what we do for statistics. We can use the database for that. We should keep in memory only the active torrents.

**Comment by mickvandijke** [comment]
During our last meeting we discussed a few potential solutions to solve this problem:

### Return the full repository of torrents without pagination

Would just return the entire torrent repository on request.

Pros:
- Works with a sharded torrent repository implementation.
- Very easy to implement (just need to remove the pagination and limit in the resource).

Cons:
- Response can get VERY big depending on the size of the torrent repository.

### Paginated and limited (current implementation)

Allows users to set a `page size` and `page number` during a request. On request it will iterate over the torrent repository and skip over `x, x = pageSize * (pageNumber - 1)` torrents and then return `page size` amount of torrents.

Pros:
- Allows for smaller responses depending on page limit.

Cons:
- Does not guarantee that the torrent pages will stay the same in between requests (the torrent repository might update).

### Paginated and limited, but with a torrent repository persistency check

Works the same as above, but also returns a unique identifier on request that represents the current state of the torrent repository. The identifier would be held in a `RwLock<Option<Identifier>>`. It gets updated whenever a user requests the resource and the `Option<Identifier>` is empty. When the torrent repository state changes in a way that the pagination would be changed as well, the `Some(Identifier)` would be dropped (if it is not already `None`).

Pros:
- Allows for smaller responses depending on page limit.
- Informs the user if the torrent pages have changed.

Cons:
- The torrent pages could be changed so often that this method might not be very useful.

### Paginated and limited and also persist insertion order of all torrent info hashes in a separate lock free set (the solution we settled on)

Add an additional struct that will keep of the insertion order of all torrents, even removed ones. Use a lock free set like [BoxCar](https://crates.io/crates/boxcar) for this struct so that we do not force all threads to content for the same write lock after announcing new torrents. Works the same as the current paginated and limited implementation for the rest, except that the order of torrents will come from this new insertion order struct instead of the torrent repository map itself. Removed torrents will be included in the JSON response, but with their value set to `null`.

Pros:
- Allows for smaller responses depending on page limit.
- Guarantees that all torrent pages stay the same because it follows the insertion order of torrents.

Cons:
- Depending on the internal struct used for the lock free insertion order struct, the last few pages of torrents could change between requests. This is because in a lock free Vec, new items are put into empty spaces near the end of the vec and not perfectly in order. This may result in some empty spots that can later be filled up.
- Uses more memory.

CC @josecelano @da2ce7

**Comment by josecelano** [comment]
Hi @WarmBeer thank you for the summary. I have a question: doesn't the `BoxCar` solution always increase the amount of memory consumed? Shouldn't we remove the "removed" torrents from time to time?

**Comment by mickvandijke** [comment]
> Hi @WarmBeer thank you for the summary. I have a question: doesn't the `BoxCar` solution always increase the amount of memory consumed? Shouldn't we remove the "removed" torrents from time to time?

Hi @josecelano ,

The memory will indeed always increase over time for every newly announced torrent. But we can not remove any info hash because it will change the torrent pagination.

**Comment by josecelano** [comment]
> > Hi @WarmBeer thank you for the summary. I have a question: doesn't the `BoxCar` solution always increase the amount of memory consumed? Shouldn't we remove the "removed" torrents from time to time?
> 
> Hi @josecelano ,
> 
> The memory will indeed always increase over time for every newly announced torrent. But we can not remove any info hash because it will change the torrent pagination.

:-/ am I missing something? @WarmBeer I guess we should "clean" the data structure from time to time or when we reach a size limit at least. I think it does not make sense to 

- [ ] https://github.com/torrust/torrust-tracker/issues/567

but introduce a new data structure that can grow indefinitely.

**Comment by mickvandijke** [comment]
> > > Hi @WarmBeer thank you for the summary. I have a question: doesn't the `BoxCar` solution always increase the amount of memory consumed? Shouldn't we remove the "removed" torrents from time to time?
> > 
> > 
> > Hi @josecelano ,
> > The memory will indeed always increase over time for every newly announced torrent. But we can not remove any info hash because it will change the torrent pagination.
> 
> :-/ am I missing something? @WarmBeer I guess we should "clean" the data structure from time to time or when we reach a size limit at least. I think it does not make sense to
> 
>     * [ ]  [Add a configurable memory limit to the torrent repository #567](https://github.com/torrust/torrust-tracker/issues/567)
> 
> 
> but introduce a new data structure that can grow indefinitely.

Hi @josecelano ,

Cleaning this data structure would defeat the whole purpose of it. We need this to persist the order of all torrents so that we can offer torrent pagination for this API endpoint `v1/torrents` without the pages changing all the time.

But yeah, it is a bit cross indeed that we now introduce a new data structure that can keep growing in memory. We could write it to disk/database instead of keeping it in memory, but that will slow down the tracker as such an operation is very slow.

I'm back to thinking that we should just drop the pagination and return the entire torrent map at once 😅. Even though this can make the response very big and slow if you have millions of torrents. I see this as the only way to reliably send all torrents back to the client without sacrificing performance in the tracker.

Perhaps we should think about why people would need this endpoint so that we can think of other solutions that would fit their needs and if returning a in the worst case big and slow response is a problem for them. We already know that it is useful for testing. 

## Use cases

### Testing

It is an easy way to test if a tracker contains all the torrents you would expect it to have.

### Scraping

External services could easily scrape all torrents from the tracker. Doing it like this is not standard protocol however.

**Comment by josecelano** [comment]
Hi @WarmBeer. I think, in the end, the `BoxCar` vector is like a database table but in memory with all the announced torrents.

Maybe the core tracker should have its own optimized in-memory data structure and that should not be directly connected to anything to avoid performance problems.

The API could be connected only to the persisted data. We could update the database directly from services. When we receive an announce request we update the core tracker but also the DB. 

If we use a data structure like this:

https://github.com/torrust/torrust-tracker/discussions/329

or even event sourcing, writes should be fast. There would be only performance problems if there were a lot of request for the API. But you have the option to disable the API.

In the future the API could be a decouple app that listens to events sent by the tracker.

cc @da2ce7

**Comment by da2ce7** [comment]
@josecelano @WarmBeer 
The box-car vector contains only a small amount of data per entry.

`BoxCar<sync::Weak<Torrent>>`

This is memory efficient.

Yes is it is a forever growing set; however only a small amount of data is kept on the stack per entry.

**Comment by josecelano** [comment]
> @josecelano @WarmBeer The box-car vector contains only a small amount of data per entry.
> 
> `BoxCar<sync::Weak<Torrent>>`
> 
> This is memory efficient.
> 
> Yes is it is a forever growing set; however only a small amount of data is kept on the stack per entry.

Hi @WarmBeer, @da2ce7 and I were discussing in today and supposing the `BoxCar<sync::Weak<Torrent>>` size is only 8 bytes.

100_000_000 torrents -> 800_000_000 bytes -> 800 MB

Our current live demo server has 1GB of RAM but 800 MB should not be a problem for a "normal" server. Besides, if you restart the service it starts from zero again.

**Comment by mickvandijke** [comment]
> BoxCar<sync::Weak<Torrent>>

Hi @josecelano , @da2ce7 ,

I just realized there is another perhaps cleaner solution. 

This is our current structure:

```rust
pub struct RepositoryDashmap {
    pub torrents: DashMap<InfoHash, Entry>,
    pub shard_priority_list: Vec<Mutex<VecDeque<InfoHash>>>,
}

pub struct Entry {
    pub peers: PeersList,
    pub completed: u32,
}

pub type PersistedTorrents = BoxCar<Weak<InfoHash>>;
```

What if we renamed the current `Entry` struct to something like `TorrentDetails` and create a new `Entry` struct like so:

```rust
...

pub struct Entry {
    // Contains the old `Entry` struct.
    pub torrent_details: Option<TorrentDetails>,
    pub time_inserted_in_nanos: u64,
}

pub struct TorrentDetails {
    pub peers: PeersList,
    pub completed: u32,
}
```

Instead of culling the entire DashMap shard entry of a torrent whenever it is the lowest priority and the TorrentRepository needs extra space because it reached the memory limit. We instead just drop `torrent_details` of a torrent entry. Only in the case that all the possible `torrent_details` have been dropped and we still need more space for a new torrent, we will actually drop entire torrent entries from the DashMap shards beginning from the last inserted. This way only the later pages of the torrents response could frequently be updated. And we still adhere to the set memory limit.

**Comment by josecelano** [comment]
Hi @WarmBeer It's not clear to me what you want to achieve. I guess you want to get rid of the secondary data structure `PersistedTorrents` keeping the "empty" torrent entry in the Dashmap. Is that right?

### Assuming you want to remove the `PersistedTorrents` type

After reading your comment the first time I thought it was a good idea but after thinking about it twice I don't think so because we would have both the API and the tracker requests hitting the same data structure. Trackers and the API are different services and they could have different load. FOr example,

- Many tracker requests, no API requests
- Many API requests, no tracker requests
- Many requests to both the trackers and API.

So the API and the trackers have to compete to write on the same data structure. We have been testing the repository only with tracker (announce) requests but not simultaneously with tracker and API requests.

In the future, both services could even be different servers if we want to scale it up.

On the other side, I guess removing the last inserted entry from the Dashmap maybe is not a fast operation. I guess you have to iterate over all items in the dashmap, right?

Finally, I don't understand your last sentence:

> This way only the later pages of the torrents response could frequently be updated

Do you mean the tracker API when you try to load the later pages of the torrent list?

- Doesn't that affect the tracker responses? If you remove the peer list from the entry.
- If you have a lot of announce requests for different torrents you would end up with empty peer list for tracker announce responses. 

But I don't see how are you going to get te ordered list of torrents for the API.

### Assuming you don't want to remove the `PersistedTorrents` type

In this case, I don't see why you want to keep the empty entry.

**Comment by mickvandijke** [comment]
Hi @josecelano ,

This is just a more efficient implementation of the `BoxCar` approach. 

### Pros
- No duplicate storage of info hashes as the torrent repository will also have the responsibility to persist the torrent order and thus we no longer need the `BoxCar` struct to persist torrents as well. Saving 20 bytes per duplicate info hash.
- The torrent repository memory limit will still be followed.

### Cons
- Every torrent will store 4 more bytes to save the insertion time in nano seconds.

### Side note
The original `BoxCar` solution has one problem. `BoxCar` can not be used as a set, so it could contain duplicate info hashes. Unless you scan it for duplicates first, but then it would be pretty slow.

> After reading your comment the first time I thought it was a good idea but after thinking about it twice I don't think so because we would have both the API and the tracker requests hitting the same data structure. Trackers and the API are different services and they could have different load. FOr example,

>    Many tracker requests, no API requests
    Many API requests, no tracker requests
    Many requests to both the trackers and API.

> So the API and the trackers have to compete to write on the same data structure. We have been testing the repository only with tracker (announce) requests but not simultaneously with tracker and API requests.

> In the future, both services could even be different servers if we want to scale it up.

Currently no API endpoint needs write access to the torrent repository as far as I'm aware. But even if we need it in the future, I do not see a problem. The dashmap implementation is a very scalable solution for concurrent write access to the torrent repository. The API would just be another interface to read and modify the torrent repository like the current UDP and HTTP tracker client interfaces are.

> On the other side, I guess removing the last inserted entry from the Dashmap maybe is not a fast operation. I guess you have to iterate over all items in the dashmap, right?

Yes this is true. I would be slower than grabbing the latest inserted info hash from a stack for example.

> Finally, I don't understand your last sentence:

>    This way only the later pages of the torrents response could frequently be updated

> Do you mean the tracker API when you try to load the later pages of the torrent list?

>    Doesn't that affect the tracker responses? If you remove the peer list from the entry.
    If you have a lot of announce requests for different torrents you would end up with empty peer list for tracker announce responses.

Yes, but that would be no different than the original `BoxCar` solution. This is a sacrifice we have to make to adhere to the optionally set memory limit. 

If the memory limit is never reached, this won't be a problem at all.

> But I don't see how are you going to get te ordered list of torrents for the API.

> Assuming you don't want to remove the PersistedTorrents type

The persistent order of torrents will be decided by the original insertion time in nano seconds that would be added to the torrent entries of the torrent repository by this proposal.

> In this case, I don't see why you want to keep the empty entry.
 
Solely to persist the torrent order for the torrent API endpoint. The entry would only contain the info hash and the insertion time.

**Comment by josecelano** [comment]
Hi @WarmBeer, thank you! I think this is a better solution because:

- I think we should not have duplicates in the BoxCar.
- This solution also has a memory limit.

> Currently no API endpoint needs write access to the torrent repository as far as I'm aware. But even if we need it in the future, I do not see a problem. The dashmap implementation is a very scalable solution for concurrent write access to the torrent repository. The API would just be another interface to read and modify the torrent repository like the current UDP and HTTP tracker client interfaces are.

I think we could add an [announce and scrape endpoints](https://github.com/torrust/torrust-tracker/discussions/708), but even without those endpoints the API needs to get the torrents. In The BoxCar solution, the API does not need to use the main repository because InfoHashes are duplicated in the BoxCar. All requests to the `/api/v1/torrents` endpoint would use the BoxCar instead of the DashMap.

Finally, I don't see how you are going to build the pagination for the `/api/v1/torrents` endpoint in the DashMap. I don't see any method to get the first 10 inserted item in the DashMap using the `time_inserted_in_nanos` field of the `Entry`.

**Comment by mickvandijke** [comment]
Hi @josecelano ,

> > Currently no API endpoint needs write access to the torrent repository as far as I'm aware. But even if we need it in the future, I do not see a problem. The dashmap implementation is a very scalable solution for concurrent write access to the torrent repository. The API would just be another interface to read and modify the torrent repository like the current UDP and HTTP tracker client interfaces are.
>
> I think we could add an [announce and scrape endpoints](https://github.com/torrust/torrust-tracker/discussions/708), but even without those endpoints the API needs to get the torrents. In The BoxCar solution, the API does not need to use the main repository because InfoHashes are duplicated in the BoxCar. All requests to the `/api/v1/torrents` endpoint would use the BoxCar instead of the DashMap.

The `/api/v1/torrents` endpoint returns more than just the info hash of a torrent. The `BoxCar` struct would only know of the info hash and would have to fetch the rest of the torrent information from the torrent repository either way. 

> Finally, I don't see how you are going to build the pagination for the `/api/v1/torrents` endpoint in the DashMap. I don't see any method to get the first 10 inserted item in the DashMap using the `time_inserted_in_nanos` field of the `Entry`.

Torrents would be sorted on the fly per API request. 

In spaghetti Rust code:

`&torrent_repository.sort(torrent.insertion_time).skip(OFFSET).take(LIMIT).collect()`

**Comment by josecelano** [comment]
Hi @WarmBeer,

> Hi @josecelano ,
> 
> > > Currently no API endpoint needs write access to the torrent repository as far as I'm aware. But even if we need it in the future, I do not see a problem. The dashmap implementation is a very scalable solution for concurrent write access to the torrent repository. The API would just be another interface to read and modify the torrent repository like the current UDP and HTTP tracker client interfaces are.
> > 
> > 
> > I think we could add an [announce and scrape endpoints](https://github.com/torrust/torrust-tracker/discussions/708), but even without those endpoints the API needs to get the torrents. In The BoxCar solution, the API does not need to use the main repository because InfoHashes are duplicated in the BoxCar. All requests to the `/api/v1/torrents` endpoint would use the BoxCar instead of the DashMap.
> 
> The `/api/v1/torrents` endpoint returns more than just the info hash of a torrent. The `BoxCar` struct would only know of the info hash and would have to fetch the rest of the torrent information from the torrent repository either way.

That's true. I didn't remember that.

> 
> > Finally, I don't see how you are going to build the pagination for the `/api/v1/torrents` endpoint in the DashMap. I don't see any method to get the first 10 inserted item in the DashMap using the `time_inserted_in_nanos` field of the `Entry`.
> 
> Torrents would be sorted on the fly per API request.
> 
> In spaghetti Rust code:
> 
> `&torrent_repository.sort(torrent.insertion_time).skip(OFFSET).take(LIMIT).collect()`

If I and ChatGTP are not wrong that means you need to:

In Rust, `DashMap` is a concurrent hash map that is designed for high performance and scalability across threads, which makes it a popular choice for multi-threaded scenarios. Unlike standard collections such as `Vec` or `HashMap`, `DashMap` does not inherently support sorting because it is primarily optimized for concurrent access, not for ordered storage.

However, if you need to sort the items stored in a `DashMap`, you can extract the items and then sort them. Here's a general approach to do so:

1. **Extract Items**: First, you need to extract all key-value pairs (or just the values or keys, depending on your requirement) from the `DashMap`.

2. **Sort**: Once you have extracted the items into a vector or another suitable data structure, you can sort them using Rust's sorting methods, like `sort` or `sort_by` for vectors.

3. **(Optional) Insert Into a Sorted Structure**: If you need to keep them sorted for later use, you might insert them into a data structure that maintains order, such as a `BTreeMap` for key-value pairs or a sorted `Vec` for values or keys.

Here's a basic example of how to do this:

```rust
use dashmap::DashMap;
use std::collections::BTreeMap;

fn main() {
    // Assume dashmap is already populated
    let dashmap = DashMap::new();
    dashmap.insert("c", 3);
    dashmap.insert("a", 1);
    dashmap.insert("b", 2);

    // Extract and sort items
    let mut items: Vec<_> = dashmap.iter().map(|entry| (*entry.key(), *entry.value())).collect();
    items.sort_by(|a, b| a.0.cmp(&b.0)); // Sorting by keys, you can sort by values or custom logic

    // Alternatively, directly insert into a BTreeMap to keep it sorted
    let mut sorted_map = BTreeMap::new();
    for (key, value) in items {
        sorted_map.insert(key, value);
    }

    // Iterating over the sorted items
    for (key, value) in sorted_map {
        println!("{}: {}", key, value);
    }
}
```

In this example:

- **Extraction**: We use `dashmap.iter()` to get an iterator over the key-value pairs, then collect them into a `Vec`.
- **Sorting**: We sort the `Vec` of tuples using `sort_by`, comparing the keys. You could adjust this to sort by values or use custom sorting logic.
- **Storing Sorted**: The sorted items are then inserted into a `BTreeMap`, which maintains sorted order based on keys.

Remember, this process of sorting is not concurrent and does not benefit from the concurrent nature of `DashMap`. If you need to frequently read sorted data in a concurrent context, consider if a different data structure or strategy might better serve your needs, such as maintaining separate sorted structures and synchronizing updates between them.
